pyop2/mpi.py

```diff
@@ -206,44 +407,62 @@ def free_comm(comm, remove=True):
             else:
                 ocomm.Delete_attr(innercomm_keyval)
             del icomm
-        if remove:
-            # Only do this if not called from free_comms.
+        try:
             dupped_comms.remove(comm)
+        except ValueError:
+            debug(f"{comm.name} is not in list of known comms, probably already freed")
+            debug(f"Known comms are {[d.name for d in dupped_comms if d != MPI.COMM_NULL]}")
         compilation_comm = get_compilation_comm(comm)
-        if compilation_comm is not None:
-            compilation_comm.Free()
+        if compilation_comm == MPI.COMM_NULL:
+            comm.Delete_attr(compilationcomm_keyval)
+        elif compilation_comm is not None:
+            free_comm(compilation_comm)
+            comm.Delete_attr(compilationcomm_keyval)
         comm.Free()
+    else:
+        warning('Attempt to free MPI_COMM_NULL')
```

Comment:
I think we want to allow this, just in case a user somehow cleans up an internal comm. The warning will help with debugging if it genuinely is a problem though.